doc/upgrade_db.1
.\" Man page for upgrade_db.1
.\" Use the following command to view man page:
.\"
.\"  tbl upgrade_db.1 | nroff -man | less
.\"
.TH UPGRADE_DB 1 "January 20, 2007"
.SH NAME
upgrade_db \- A tool to upgrade Emdros databases
.SH SYNOPSIS
\fBupgrade_db\fR [options] database-name
.br
.SH DESCRIPTION

\fBupgrade_db\fR is a tool for upgrading Emdros databases created with
Emdros version 1.1.7 or earlier to databases that work with Emdros
version 1.1.8 and later.

.br
For databases created with Emdros prior to version 1.1.0, the program
does the following: For each object type OT in the database, it:

1) Adds a new column, \fIis_first\fR, to the \fIOT_monad_ms\fR table.

2) Adds some indices on \fIOT_monad_ms\fR and \fIOT_objects\fR (see EMdFDB::createIndicesOnOTObjects and 
EMdFDB::CreateIndicesOnOTMonadMs).

3) If using PostgreSQL, sets all \fIis_first\fR values in \fIOT_monad_ms\fR to 'N'.

4) Finds out which \fIis_first\fR values should be set to 'Y'.

5) Sets these to 'Y'.  Each dot represents 500 rows.

6) Drops table \fIOT_monad_ds\fR.

After having dealt with all object types, the program creates two new
tables, `sequence_0' and `sequence_1', and moves the values from the
old `sequences' table into these new tables.  It then drops the old
`sequences' table.

For databases created with Emdros versions 1.1.0 to 1.1.4, the program
does the following:

1) Creates the table \fImax_m\fR

2) Sets the values in this table to the maximum monad found in the
database.

For databases created with Emdros versions 1.1.5 to 1.1.7, the program
does the following:

1) Creates the tables/sequences \fIsequence_118_0\fR and
\fIsequence_118_1\fR (tables on MySQL, sequences on PostgreSQL).

2) Moves the values from the old \fIsequence_0\fR and \fIsequence_1\fR
tables to the new tables/sequences.

3) Drops the old \fIsequence_0\fR and \fIsequence_1\fR tables.

See the source in EMdF/emdfdb.cpp:EMdFDB::upgradeDatabase for the full
story.

.SH OPTIONS
\fBupgrade_db\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-h\ ,\ \-\-host \ hostname 
set db back-end hostname to connect to (default is 'localhost'). Has
no effect on SQLite.
.TP
.BI \-u\ ,\ \-\-user \ user
set database user to connect as (default is 'emdf'). Has no effect on
SQLite.
.TP
.BI \-p\ ,\ \-\-password \ password
set password to use for the database user.  Has no effect on SQLite,
unless you have an encryption-enabled SQLite, in which case this gets
passed as the key.
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".




.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not upgrade"
.TP
.BR 5 " Database error"
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.


- *** Version 3.6.1 *** 

Release notes for Emdros 3.6.1
==============================

Emdros version 3.6.1 is a maintenance release with the following
summary of changes since 3.6.0:

- The MQL query language has been extended with the ability to compare
  features with type LIST OF INTEGER, LIST OF ID_D, and LIST OF
  <enum-type> with an explicit list (including the empty list) using
  the "=" comparison operator.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause_atom
     daughters = () // Test for empty list
     OR
     daughters = (12345,67891) // Test for explicit, non-empty list
  ]
  GO

- Detection of PHP 5 and PHP 7 on macOS now works correctly.

- Compileability on Solaris 10 with Sun's CC C++ compiler has been
  restored.

- Compatibility with wxWidget 2.8.12 has been restored.


Enjoy!


Ulrik Sandborg-Petersen



- *** Version 3.6.0 *** 

Release notes for Emdros 3.6.0
==============================

The release notes below cover some of the important changes in Emdros
since 3.5.0.

- SQLite2 support has been removed, as version 2 of SQLite has not
  been updated by the SQLite authors for many, many years.  Also,
  SQLite2 support in Emdros has been been deprecated for some time.

- The MQL subsystem has a better query planner, which is able to
  choose the "Innermost First" strategy, rather than the default
  "Outermost First" strategy, under many more circumstances.

- The Emdros MySQL backend now has support for transactions, if
  available in the MySQL server.

- Assorted performance improvements to the following subsystems:

  - XML parsing (renderobjects, for example)
  - MQL

- Node.JS support has been removed.  The reason is that the latest
  SWIG version (3.0.12) does not support the latest Node.  Also, the
  Node.JS bindings never really worked.

- PCRE has been upgraded.

- SQLite3 has been upgraded.

- JSON output now escapes string characters correctly.

- Compiling Linux .rpm packages works on recent systems (CentOS 7,
  Fedora 28).

- Compiling Linux .deb packages works on recent systems (Ubuntu 16.04,
  Ubuntu 18.04, Debian 9.4).

- Building on Windows supports the latest Visual Studio.

- Lots of compiler warnings have been removed.

- the harvest template language has new constructs:

  - {{ inccounter <string> var <string> }}
  - {{ deccounter <string> var <string> }}

- The Full Text Search FTS3 subsystem is now able to obey
  largest_proximity, and to use * and ? in the query terms.

- A new program was added, bptdetach(1), which enables detaching
  "payloads" from BPT databases.

- The BPT backend (if available -- ask Ulrik for the source code
  licensing terms if you are interested) now supports LZMA compression
  of payloads.

- Source code using wxWidgets has been updated to support wxWidgets
  2.8, 3.0, and 3.1.

- The Emdros PostgreSQL backend now adds "TEMPLATE template0" when
  doing CREATE DATABASE.  This is so as to avoid using template1,
  which may conflict with the encoding used (and does so on Debian
  9.4).

- The amalgamation output has been split into several files.  This is
  because the Android NDK MIPS target cannot compile the old,
  monolithic amalgamation files.



- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries

src/upgrade_db.cpp
/*
 * Upgrade_db.cpp
 *
 *
 *
 * Tool to upgrade an Emdros database created with version 1.1.4 or 
 * earlier to one that works with Emdros release 1.1.8 and later.
 *
 * Run it with the name of the database to upgrade as the last
 * argument.
 * 
 * Supports -h, -u, and -p arguments (see output from --help).
 *
 * For example:
 *    ./upgrade_db book
 * 
 *    ./upgrade_db --host myhost.somedomain.com mydb
 * 
 *
 *
 * Created: 11/11-2001
 * Last update: 5/11-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : We could not upgrade
 * 5 : Database error
 * */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */



#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc

#include <monads.h>
#include <cstdio>
#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <emdfdb.h>
#include <table.h>
#include <emdros-lconfig.h>
#include <emdf_output.h>
#include <emdros_environment.h>
#include <mql_execution_environment.h>
#include <encryption.h>
#include <opt.h>
#include <messages.h>





void show_version(std::ostream& ostr)
{
	ostr << "upgrade_db from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   upgrade_db [options] databasename " << std::endl;
	ostr << "   upgrade_db --help" << std::endl;
	ostr << "   upgrade_db (-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	printUsageStandardArguments(ostr);
	ostr << "DEFAULTS:" << std::endl;
	printUsageDefaultsOfStandardArguments(ostr);
}

bool show_warning(void)
{
	std::cout << std::endl;
	std::cout << "upgrade_db:" << std::endl;
	std::cout << "  This program upgrades an Emdros database created with a version" << std::endl;
	std::cout << "  of Emdros prior to 1.1.8 to a database that can be used with Emdros" << std::endl;
	std::cout << "  version 1.1.8 and later, but only to a database version in the 1.1-series." << std::endl;
	std::cout << std::endl;
	std::cout << "  WARNING: This software may make a lot of changes to the database." << std::endl;
	std::cout << "           It is recommended that you BACKUP YOUR DATABASE before proceeding." << std::endl;
	std::cout << std::endl;
	std::cout << "  Are you sure you want to proceed? (y/n)" << std::endl;
	char c;
	c = getchar();
	return c == 'Y' || c == 'y';
}

int main(int argc, char* argv[])
{
	// Declare option-strings
	std::string database_to_upgrade;
	std::string password;
	std::string hostname;
	std::string user;
	eCharsets charset = kCSISO_8859_1;
	eBackendKind backend_kind;

	addStandardArguments(false // Do not add -e | --encoding
			     );

	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  hostname,
					  user,
					  password,
					  backend_kind,
					  charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		// -V | --version
		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}
		if (surplus_arguments.size() == 0) {
			print_usage(std::cerr);
			std::cerr << "ERROR: You must specify a database to upgrade on the command line." << std::endl;
			return 1;
		} else if (surplus_arguments.size() == 1) {
			database_to_upgrade = *surplus_arguments.begin();
		} else {
			std::cerr << "Error: You can only specify one database o upgrade. You specified: '" << joinList("' and '", surplus_arguments) << "'." << std::endl;
			return 1;
		}

		if (database_to_upgrade == "") {
			print_usage(std::cerr);
			std::cerr << "ERROR: Database name to upgrade must not be the empty string." << std::endl;
			return 1;
		}
	}

	// Make connection
	EmdrosEnv *pEE = new EmdrosEnv(kOKConsole, charset,
				       hostname,
				       user, password,
				       database_to_upgrade,
				       backend_kind);

	EMdFDB *pDB;
	pDB = (pEE->getMQLEE())->pDB;

	// Zero-fill password
	zeroFillString(password);

	int nResult = 3;
	try {

		// Check that we connected
		if (!pDB->connectionOk()) {
			emdrosMessageConnectionNotOK(&std::cerr, backend_kind);
			delete pEE;
			return 2;
		}

		// Check whether we need upgrading
		std::string strCurrentDBVersion;
		bool bNeedsUpgrading;
		bool bCanUpgrade;
		int nCurrentVersion;
		int nVersionWhichWeCanUpgradeTo;
		std::string strVersionWhichWeCanUpgradeTo;
		if (!pDB->getCurrentDatabaseVersion(database_to_upgrade, 
						    "", // Right now, we only upgrade PostgreSQL and MySQL databases, so we just pass an empty key.
						    strCurrentDBVersion, 
						    nCurrentVersion, 
						    nVersionWhichWeCanUpgradeTo, 
						    strVersionWhichWeCanUpgradeTo, 
						    bNeedsUpgrading, 
						    bCanUpgrade)) {
			std::cerr << "ERROR: Database error..." << std::endl;
			std::cerr << pDB->errorMessage() << std::endl; 
			std::cerr << "Could not determine current database version. Aborting without modifying DB." << std::endl;
			delete pEE;
			return 5;
		}
    
		// Show current version
		std::cout << "Current database version:" << std::endl;
		std::cout << "    *** " << strCurrentDBVersion << " ***" << std::endl;

		// Check to see whether we should upgrade
		if (!bNeedsUpgrading) {
			std::cout << "The database is compatible with the current version of Emdros," << std::endl
				  << "and doesn't need upgrading. Aborting without modifying database." << std::endl;
			delete pEE;
			return 0;
		}

		if (!bCanUpgrade) {
			std::cout << "ERROR: Cannot upgrade the database." << std::endl;
			std::cout << "Exiting without doing anything to the database." << std::endl << std::endl;
			std::cout << "This most likely means that the upgrade_db program does not provide an" << std::endl;
			std::cout << "upgrade path from the version of the database to the current version of" << std::endl;
			std::cout << "Emdros." << std::endl << std::endl;
			std::cout << "You may be able to use the mqldump program to dump the database and reload it" << std::endl;
			std::cout << "with the current version of Emdros." << std::endl;
			delete pEE;
			return 4;
		}

		std::cout << "We can upgrade the DB to the following version:" << std::endl;
		std::cout << "    *** " << strVersionWhichWeCanUpgradeTo << " ***" << std::endl;
		if (nVersionWhichWeCanUpgradeTo != nCurrentVersion) {
			std::cout << "INFORMATION:" << std::endl;
			std::cout << "    We CANNOT upgrade directly to the current version." << std::endl;
			if (nVersionWhichWeCanUpgradeTo == EMDROS_VERSION_1_1_8_TO_1_1_20) {
				std::cout << "    It looks like you can only upgrade to a version compatible with 1.1.8." << std::endl
					  << "    This is actually good news, since mqldump from the 1.1-series works with" << std::endl
					  << "    databases from 1.1.8 and upwards.  You can then use a 1.1-series mqldump" << std::endl
					  << "    to dump the database and reimport it in the new series.  Just use an" << std::endl
					  << "    mqldump from the 1.1-series.  mqldump was introduced in 1.1.11." << std::endl;
			} else {
				std::cout << "    You may be able to use mqldump to dump the database and import it again" << std::endl;
				std::cout << "    with the current version.  Use an mqldump that matches the version of" << std::endl;
				std::cout << "    the database." << std::endl;
			}
		}

		// Show warning
		if (!show_warning()) {
			std::cout << "OK, aborting without modifying DB." << std::endl;
			delete pEE;
			return 0;
		}


		// Show our intention
		std::cout << "Upgrading database to version 1.1.8 and beyond..." << std::endl;

		if (pDB->upgradeDatabase(database_to_upgrade, nCurrentVersion)) {
			nResult = 0;
		} else {
			nResult = 5;
			std::cerr << "ERROR: Database error..." << std::endl;
			std::cerr << pDB->errorMessage() << std::endl; 
		}
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Clean up
	delete pEE;

	if (nResult == 0) {
		std::cout << "Database " << database_to_upgrade << " was successfully upgraded." << std::endl;
	} else {
		std::cerr << "An error occurred while upgrading database " << database_to_upgrade << "." << std::endl;
		std::cerr << "Reason: ";
		if (nResult == 3) {
			std::cerr << "An exception occurred. (See above.)" << std::endl;
		} else if (nResult == 5) {
			std::cerr << "A database error occurred. (See above.)" << std::endl;
		}
		std::cerr << "Database was not successfully upgraded." << std::endl;
		std::cerr << "Try restoring from a backup and trying once more." << std::endl;
		std::cerr << "If this does not work, please contact the Emdros maintainer," << std::endl;
		std::cerr << "Ulrik Petersen, at ulrikp@users.sourceforge.net with a problem description." << std::endl;
	}
	return nResult;
}
